riscv/proc_trace.cc

```diff
@@ -0,0 +1,325 @@
+      m_csr_load_data(0), m_csr_size(0)
+{
+}
+
+void proc_trace_t::open_i_trace(const char *instruction_trace_file_name) {
+    m_itrace_fd = fopen(instruction_trace_file_name, "w");
+    if (!m_itrace_fd)
+        fprintf(stderr, "Failed to open instruction trace file %s: %s",
+                instruction_trace_file_name, strerror(errno));
+    else
+        fprintf(m_itrace_fd, "VALID,ADDRESS,INSN,PRIVILEGE,EXCEPTION,ECAUSE,TVAL,INTERRUPT\n");
+}
+
+/*
+  Data trace of load/store is complicated by the fact that spike has an MMU and the default
+  is not to perform memory trace. It is also complicated by the fact that the memtracer
+  in spike does not trace I/O regions (I assume because it was designed for cache tracing and
+  I/O would not be cached).
+  IMPORTANT NOTE: the memtracer MUST be registered even though it is not used for producing
+  the trace data. The MMU changes it's behaviour when a memtracer is registered and this is
+  required.
+*/
+void proc_trace_t::open_d_trace(const char *data_trace_file_name, bool debug) {
+    m_data_trace_debug = debug;
+    if (m_data_trace_debug)
+        fprintf(stderr, "Warning: data trace debug switched on - data trace files will not be usable\n");
+    m_dtrace_fd = fopen(data_trace_file_name, "w");
+    if (!m_dtrace_fd)
+        fprintf(stderr, "Failed to open data trace file %s: %s",
+                data_trace_file_name, strerror(errno));
+    else
+        fprintf(m_dtrace_fd, "DRETIRE,DTYPE,DADDR,DSIZE,DATA\n");
+}
+
+void proc_trace_t::step(void) {
+    if (!m_itrace_fd)
+        return;
+
+    if (m_is_first_step) {
+        m_is_first_step = false;
+        return;
+    }
+
+    reg_t addr = m_addr;
+    reg_t tval = m_tval;
+    if (m_is_32bit_isa) {
+        addr &= 0xffffffff;
+        tval &= 0xffffffff;
+    }
+    fprintf(m_itrace_fd, "1,%lx,%lx,%x,%x,%lx,%lx,%x\n",
+            addr, m_insn_binary, m_prv, m_ex, m_ex_cause, tval, m_interrupt);
+    fflush(m_itrace_fd);
+
+    m_ex = 0;
+    m_ex_cause = 0;
+    m_tval = 0;
+    m_interrupt = 0;
+}
+
+void proc_trace_t::set_addr(reg_t addr) {
+    m_addr = addr;
+}
+
+void proc_trace_t::set_insn(insn_t insn, uint64_t binary) {
+    m_has_insn = true;
+    m_insn = insn;
+    m_insn_binary = binary;
+    if (m_dtrace_fd) {
+        m_dtype = DTYPE_INVALID;
+        m_recorded_dtype = DTYPE_INVALID;
+        m_csr_which = CSR_INVALID;
+        m_csr_write = false;
+        m_csr_load_data = 0;
+        m_csr_size = 0;
+
+        bool is_compact = (binary & 0x3) != 0x3;
+        m_is_amo = false;
+        if (is_compact) {
+            // Not interested in load immediate
+            uint64_t inst_15_13 = (binary >> 13) & ((insn_bits_t(1) << 3) - 1);
+            uint64_t inst_1_0 = binary & 0x3;
+            m_is_load = (((inst_1_0 == 0) || (inst_1_0 == 2)) && (inst_15_13 > 0) && (inst_15_13 < 4));
+            m_is_store = ((inst_1_0 == 0) || (inst_1_0 == 2)) & (inst_15_13 > 4);
```

Comment:
This is not robust, as new load & store opcodes might be added to the ISA at any time, or in a custom fork of Spike, and they would not know to update this cryptic decoder here.

Spike already has some way of decoding whether an opcode is a load/store, though it's not clear to me why this is needed here, since the act of loading or storing will call other methods in this class. That would be a much more reliable way of determining that an instruction did a load or store.

Reply:
Could you please point me at the way of deciding whether an opcode is a load or store? I had looked in decode.h, but couldn't see any obvious way of doing it. I agree that I really don't want this partial disassembly in this code.
Thanks.

Reply:
What I meant is, if you see `record_load()`, `record_store()`, or `record_amo()` called, then you know it's a load/store. If not, then it's not.

Reply:
I have removed the disassembly code. The check for whether it's an atomic instruction is required because otherwise you get spurious load/stores being recorded. I didn't have a test for this in the subset of tests I have imported, so I've added a test which requires this check.
I also need to be able to determine what type of instruction it is for the data trace of CSR and atomics so I've now done that using the constants defined in encoding.h.

Reply:
Can you explain what you mean by "spurious load/stores"? An AMO is both a load and a store. What's supposed to get recorded in the trace in that case? I wonder if a solution for #873 would be able to help here.

Reply:
The data trace spec differentiates AMO from other load/store operations. The problem I am trying to solve here is, to make sure that the load/store operations are not recorded in the trace, if the instruction is actually an AMO.
Regarding #873, it may be that a solution to that helps here.

Reply:
Table 4.12 in the trace spec (https://github.com/riscv-non-isa/riscv-trace-spec/blob/main/riscv-trace-spec.pdf) gives a list of the types of data access that are to be recorded as part of the data trace.